hs_collection_resource/management/commands/fix_collection_version_missing_ispartof.py

```diff
@@ -17,19 +18,85 @@ def add_arguments(self, parser):
+                    print(f"{n}/{col_count}:For collection:{collection}, {current_site}/resource/{collection.short_id}")
+                    print(f"IsPartOf relation missing from {res}, {current_site}/resource/{res.short_id}")
                     if dry_run:
                         print("SKIPPING creating IsPartOf relation because dry_run")
                     else:
                         print("Creating IsPartOf relation...")
                         res.metadata.create_element('relation',
                                                     type=RelationTypes.isPartOf,
                                                     value=collection.get_citation())
+                    fixed = fixed + 1
+                res_citation = res.get_citation()
+                if not collection.metadata.relations.filter(type=RelationTypes.hasPart, value=res_citation).exists():
+                    print(f"hasPart relation missing from collection: "
+                          f"{collection}, {current_site}/resource/{collection.short_id}")
+                    if dry_run:
+                        print("SKIPPING creating hasPart relation because dry_run")
+                    else:
+                        collection.metadata.create_element('relation',
+                                                           type=RelationTypes.hasPart,
+                                                           value=res_citation)
+                    fixed = fixed + 1
+        print(f"{fixed} metadata relations were missing")
+
+        print('-' * 100)
+        print("2. Remove dangling isPart metadata from resources that are not included in collections")
+        resources = BaseResource.objects.all()
+        res_count = resources.count()
+        print(f"{res_count} resources to check")
+        i = 0
+        fixed = 0
+        for res in resources:
+            i = i + 1
+            isPartOf_relations = None
+            if res.metadata:
+                isPartOf_relations = res.metadata.relations.filter(type=RelationTypes.isPartOf).all()
+            if not isPartOf_relations:
+                continue
+            for rel in isPartOf_relations:
+                # get the collection object from the relation...
+                try:
+                    citation = rel.value
+                    if "doi.org" in citation:
+                        id = rel.value.split("/hs.")[1][:32]
+                    else:
+                        id = rel.value.split("/resource/")[1][:32]
+                    col = BaseResource.objects.get(short_id=id)
+                except Exception as e:
+                    message = f"Unable to parse id from citation: {rel.value}, {e}"
+                    print(message)
+                    errors.append(message)
+                    continue
+                haspart_relations = col.metadata.relations.filter(type=RelationTypes.hasPart).all()
+                citation = res.get_citation()
```

Comment:
may this line we should move up before line 72 to reduce queries.

Reply:
good call thanks